### PR TITLE
OAK-10535: Clean up old revisions in a document

### PR DESCRIPTION
**Pull request by Joscorbe**
Introduce a way to safely detect and delete old revisions in a document. This could be useful to cleanup documents that sometimes grows above the supported size in MongoDB (16MB).

This is still WIP and needs to be integrated into the DetailedGC logic.

This PR should supersede #1191

### CODE REVIEW

File: oak-run/src/main/groovy/org/apache/jackrabbit/oak/console/GroovyConsole.groovy
Comment (stefan-egli) [inline oak-run/src/main/groovy/org/apache/jackrabbit/oak/console/GroovyConsole.groovy]: You might already be aware of this, but just to be sure : do we need a new RevisionCommand when we have the RevisionsCommand?

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (stefan-egli) [inline oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: We might be able to handle this under the existing `collect` command - that one currently runs classic GC, but you can get it to do detailed GC as well, with the corresponding parameter.

(What's then missing is the ability to restrict it to specifically only a path)

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java]: Is it required to have the DocumentStore passed here? The `DocumentNodeStore.getDocumentStore()` would provide it too - but that way it would be avoided that for some reason the two could not link to the same...

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java]:
```suggestion
```

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java:37]: Some javadoc to describe what this class is about would be good to have.

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (stefan-egli) [inline oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: As was discussed offline, we'd need to change this into a way which calls GC in general (or if that is unfeasible at least call DetailedGC in general).

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleanerTest.java
Comment (stefan-egli) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleanerTest.java]: This shows up as deprecated. Would using the suggested alternative (`MockitoAnnotations.openMocks` - plus a later `close()`) work?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java]: Is there a reason this is not `private` ?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java]: Is there a reason this is not `private` ?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java:161]: While these are internal (presumably `private`) classes, so no expected interference from elsewhere, and given these are not value or holder but more worker type of classes, it might be worth while to avoid direct field access between each other (and instead go via getter methods)

**Comment by stefan-egli** [comment]
It's generally speaking a bit difficult to review as the code doesn't compile at this point (therefore eg you can't debug through it, to understand it better). As I understand this is soon going to be fixed - so I'll wait with reviewing actual functionality for now.

File: oak-run/src/main/groovy/org/apache/jackrabbit/oak/console/GroovyConsole.groovy
Comment (Joscorbe) [reply oak-run/src/main/groovy/org/apache/jackrabbit/oak/console/GroovyConsole.groovy]: The intention of this was to be able to run from the interactive console. We would probably don't this anymore, since that console is read-only anyway.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java
Comment (Joscorbe) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java]: In the initial version this would receive a read-only DS and the write one. It is now obsolete since I don't think we are going to run this from oak-run console, but from the revisions command.

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (Joscorbe) [reply oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: I agree we should at some point combine some of the commands into the same one.
For now, I have changed this to `pathCleanup`. But I think a better name can be found.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollectorIT.java
Comment (Joscorbe) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollectorIT.java:1]: I haven't done any changes in this file.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java
Comment (Joscorbe) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java]: Left over from initial testing. I have made it private.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java
Comment (Joscorbe) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java]: Left over from initial testing. I have made it private.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleanerTest.java
Comment (Joscorbe) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleanerTest.java]: Done

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (Joscorbe) [reply oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: This now calls the general detailedGC

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java
Comment (Joscorbe) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java:37]: Done. The class is a prototype, probably we will never use it in production, as deep cleaning in-between checkpointed revisions is probably too much.

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (rishabhdaim) [inline oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: please remove these un-used imports.

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (rishabhdaim) [inline oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: Not used inside code. Please remove this.

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (rishabhdaim) [inline oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: we are removing whole detailed garbage here. I would change this description to reflect that it is performing detailedGC on given path.

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (rishabhdaim) [inline oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: Returning `UpdateOp` from detailedGC and handling deletion here is not a good idea, cause we might need to duplicate the code written in `DetailedGC.removeGarbage()` OR we might miss something.

In this case, we are missing orphan nodes cause they are handled differently in `detailedGC.removeGarbage()`.
I would delegate the collection & removal of garbage to DetailedGC itself.

**Review by rishabhdaim — CHANGES_REQUESTED** [review]

**Comment by Joscorbe** [comment]
I have squashed the WIP commits, sorry for the forced-PR.

**Comment by rishabhdaim** [comment]
Could you please revert `VersionGarbageCollectorIT` from your PR, it looks like you haven't made any changes in this file and it is reverting changes done in this [PR](https://github.com/apache/jackrabbit-oak/pull/1332/files#diff-b4378c829e2a8761f08cd1cb460bbc3bec86b74f379243bdea64de2bfa27a107)

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java
Comment (Joscorbe) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java:161]: Done

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (Joscorbe) [reply oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: Done

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (Joscorbe) [reply oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: Done

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (stefan-egli) [inline oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: (where) Is this still used?

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (stefan-egli) [inline oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]:
```suggestion
            "  info        give information about the revisions state without performing",
```
(minor: indentation)

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (stefan-egli) [reply oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: alternatively I was wondering if the restriction to a single path could not also be a further command line parameter of `detailedGC` mode? (in which case we wouldn't need a new mode just make `detailedGC` more fancy)

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (stefan-egli) [inline oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: Is the idea to keep this as a reminder for a future change (or could it be removed now)?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java]: Would suggest to either rename to clarify that this is about old and between-checkpoint-but-not-younger-than-maxRevisionAgeMillis revisions - or to clarify that in a javadoc.

Now, given the old-revision-gc part competes with OAK-10688 we should execute only one of the two, not both. One idea could be to add an static boolean (not a system property, just a plain boolean) that lets you run DetailedGC in either between-checkpoint mode (OAK-10535) or keep-1-revision mode (OAK-10688). Then tests can control that flag at their discretion, we can compare the two and find bugs in the two, but ultimately decide on which one we want.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java]: As an alternative idea to the 2 steps suggested (`collectGarbageOnDocument` then removing the garbage in `RevisionsCommand`) which @rishabhdaim also commented on : what about focusing on executing the [core part](https://github.com/apache/jackrabbit-oak/blob/cf870b532b107b0c243e1d0180f1213d3f6e9197/oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java#L736-L782) of `collectDetailedGarbage` in 2 different modes : the current full one, and a new `onPath` one. Eg that would imply refactoring the existing `collectDetailedGarbage`'s inner part out and then more or less directly invoking that on a particular path (perhaps have a small facade method that facilitates that).

The problem I think is indeed, as @rishabhdaim  [already mentioned](https://github.com/apache/jackrabbit-oak/pull/1328/files#r1511381906), that some parts might be missed, if we have a second variant of `removeGarbage` (for example we wouldn't do the embedded verification). Of course doing it "the real way" does require some refactoring, but I think it could be done via that inner loop mentioned. Wdyt?

Comment (Joscorbe) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java]: I'm working on this approach, will commit the changes soon.

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (Joscorbe) [reply oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: Working on this, but I'm refactoring the `collectDetailedGarbage` to run all the logic on a single document. That way I think we will also improve the readability of that code.

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (Joscorbe) [reply oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: That's possible, yes, and would be cleaner IMO. Since the `--path` is already a separated argument we can simply apply it to this command.
My only concern with this approach is, if the `path` argument is missing, it could end up running the DetailedGC in all the repository.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java
Comment (Joscorbe) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java]: After some work in that area, I'm abandoning this approach. I initially thought it would be a good idea because that method is already quite long, so splitting it could improve its readability. However, the _real core_ part would be [just these lines](https://github.com/apache/jackrabbit-oak/blob/cf870b532b107b0c243e1d0180f1213d3f6e9197/oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java#L752-L756), as the rest of the code is essentially to keep the progress of the iterators.

So I simply [added this part](https://github.com/apache/jackrabbit-oak/blob/cf870b532b107b0c243e1d0180f1213d3f6e9197/oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java#L768-L772) to the facade method I already added.

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (Joscorbe) [reply oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: Fixed

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (Joscorbe) [reply oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: No, some leftover from old code.

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (Joscorbe) [reply oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: No, it's removed now.

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (stefan-egli) [reply oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: In theory if someone runs the command they should be aware of what it does - it's an admin command after all.

I guess we're doing it this careful because the new code isn't tested well enough yet.

Could we then - during this testing phase - demand that you either specify `--path` or `--entirerepo` (or something like this)? Then eventually make `--entirerepo` the default and not demand it anymore?

**Review by rishabhdaim — APPROVED** [review]

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
Comment (Joscorbe) [reply oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java]: That makes me a little less nervous. I have added that option for now.

**Review by stefan-egli — APPROVED** [review]
+1 for merging to the feature branch. Before merging to trunk we should still introduce a flag as to which "old revision cleanup" variant we'll use - but that shouldn't block this PR as of now.

**Pull request merged**
